README.md
# NeoGAF-ThreadSummarizer-Extension

This Branch is WIP, not supposed to work and is volatile. Please download the master branch for your own work.
Cloned on Work Computer

Current known problems with the branch:
Extension is working as intended on version 1.7.0.0

This extension provides features that aid browsing of NeoGAF forums.

Copyright (c) 2015 hesido.com

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgement in the product documentation would be
   appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.


Changelog.txt
NeoGAF thread summarizer shifts through thread pages starting from the page it was called on, and then allows you to view most quoted posts with a threshold you can set. Once a thread is analysed, you can display posts with different quote threshold values. You can also track your own or any user's posts, and replies to those posts. Pages are cached so you can switch between pages really fast. V1.7 introduces live thread functionality, you are notified for new posts in the thread.
-----
Version 1.7.0.2 Change log
Further SSL related fixes

Version 1.7.0.1 Change log
Apapted to SSL change on NeoGAF

Version 1.7.0.0 Change log
New posts are now notified on a thread after analysis (optional, default on)

Version 1.6.1.3 Change log
Preparing for GIT, minor bug fixes and code clean up.

Version 1.6.1.2 Change log
Added populate pages in quick settings.

Version 1.6.1.1 Change log
After an analysis is complete, the pages are now automatically populated for quote chains. This means you don't have to switch to quoted only view, and you can go on reading the threads as usual and still track conversations.

Version 1.5.1 Change log
*Fixed an issue with animation causing memory leak
*Major new feature: Recursive thread view.  Now you can list posts that quoted posts that quoted posts .. that quoted posts. This allows you to follow a back-and-forth replies or posts that cause long chains of replies.
*Manually refresh page caches. Now you can manually update the cached pages to reflect latest edits / posts. This is also paving the way for a "live thread view", which will give the extension a paginated auto updating live view of the page.

Version 1.4.1 Change log
*This is a bug fix update.
*Fixed a bug that was preventing the last page to be updated in cache
*Turning off "cache analysed pages" setting immediately deletes all page download caches. If the extension fails to update the pages of already analysed pages, you can turn this on and off to get rid of caches.

Version 1.4.0 Change log
*There's now an advanced options page. Linked from extensions page in chrome.
*Cached pages and analysis data is automatically removed after a time of inactivity. The amount of time can be set in options page. This reduces memory consumption without the user explicitly clearing the cache.
*The analysis cache and page caches are now separate entities, so you can have cached pages without analysis cache and vice versa. This will allow future enhancements to GAF browsing.
*Re-analysis of threads are much faster as the extension now uses existing page caches except for the last page.

Version 1.3.4.1 Change log
*Fixed by-design behavior that counts the separated snippets of the same quoted post to be counted multiple times.
*Now you can re-do the analysis on top of a previous analysis without having to clear thread cache. This allows you to re-analyze last page along with the added post and pages and update the thread cache without a full analysis. Page already analysed error is no longer thrown. You can start analysis from already analysed pages to update the thread cache to the latest version of the posts and page.
*Some CSS style fixes

Version 1.3.3.1 Change log
*Slightly less memory usage when using cache function
*Changed the UI a bit
*Now the quoter post lists are collapsible
*Options are saved and synced to cloud so you don't have to re-enter settings every time browser is closed of the extension is updated.

Version 1.3.2.2 Changle log:
*Fixed a minor bug introduced in 1.3.2 that prevented correct top-quoted sort.

Version 1.3.2 Change log:
*Quoted and quoting posts are now displayed from cache if cache pages option is turned on.

Version 1.3.1 Change log:
*Analysed pages are now also cached and loaded from memory for navigating the thread (optional)
*You no longer have to reload page to change the display results settings.
*Lots of back-end changes to be better future-proof

Version 1.2.2 Change log:
*All loading and analysis now done in background so you can continue reading the page you are on.
*UI enhancements
*Abort analysis button now should work properly.
*Extension now obeys your preferred domain prefix for neogaf.com.

Version 1.1 Change log:
*It's now possible to track posts by specific users, this allows you to view replies to your post or keep track of your favorite posters.
*The UI remembers your latest settings during a session.

Version 1.0 Change log:
*The UI was slightly changed to make more sense.
*The posts are now loaded in correct order.
*Added the option to sort posts by the quote popularity (default is chronological.)
*Better error handling


//done: Cached navigation
//done: remove background.window object references in popup.js
//done: click handlers no longer to use closures
//done: Re-display results without re-load
//done: add basic cache settings on popup

//done: aggressive removal of cached pages to reduce mem usage (remedy the increased mem usage introduced by "use cached pages for post retrieval")
//done: settings are stored, synced
//done: fix behaviour that causes separate quotes of the same post in the same quoting post to be counted multiple times
//done: better behaviour when re-analysing already analyzed threads
//done: time limit assigned for auto removal of cached pages
//done: separate options page
//done: optionally use cache when re-analyzing (except last page analyzed)
//done: infinite recursive post display
//done: collapse quoters properly in recursive fashion
//done: fix: re-analysis broken after recursive post display
//done: fix: re-displaying analysis results is buggy
//done: manual refresh page to update caches
//done: auto refresh last page to notify of new posts
//change: Now, thread analysis data is kept until all tabs of the thread is closed, where they are removed after a user-defined time. Old versions looked for navigational activity to keep the thread data alive.


options.js
// NeoGAF Thread Summarizer Copyright (c) 2015 hesido.com
"use strict";

var cacheset,
usecache,
cachelim,
analimit,
statusms,
usecaclb,
pagemodi,
onpageac,
autorefr,
usecachesetting,
receivedsettings = false,
settings = {
	settingsVersion: 1.3,
	cachepages: true,
	usecacheforanalysis: true,
	pagecachetimelimit: 15,
	analysiscachetimelimit: 15,
	populatepages: true,
	onpageactions: true,
	autorefreshevery: 1,
}; //set defaults first without waiting for the local storage callback just in case;


function save_options() {
	if (!settings) return;
	settings = {
		settingsVersion: settings.settingsVersion,
		cachepages : cacheset.checked,
		usecacheforanalysis : usecachesetting, //this setting will not be directly tied to the on-screen checked status.
		pagecachetimelimit : cachelim.value || settings.pagecachetimelimit,
		analysiscachetimelimit: analimit.value || settings.analysiscachetimelimit,
		populatepages: pagemodi.checked,
		onpageactions: onpageac.checked,
		autorefreshevery: autorefr.value
	};
	chrome.storage.sync.set(settings, function () {
		// Update status to let user know options were saved.
		chrome.runtime.sendMessage(null, {
			action : "receiveSettings",
			settings : settings,
			saved : true
		});
		statusms.textContent = 'Status: Saved settings.';
		setTimeout(function () {
			statusms.textContent = 'Status: Waiting for changes.';
		}, 500);
	});
}

// Restores select box and checkbox state using the preferences
// stored in chrome.storage.
function restore_options() {
	cacheset = document.getElementById('cachepage');
	usecache = document.getElementById('usecachepages');
	cachelim = document.getElementById('pagecachetimelimit');
	analimit = document.getElementById('analysiscachetimelimit');
	statusms = document.getElementById('status');
	usecaclb = document.getElementById('usecachepages_label');
	pagemodi = document.getElementById('dopagemod');
	onpageac = document.getElementById('onpageactions');
	autorefr = document.getElementById('autorefreshlastpage');

	chrome.storage.sync.get(
		settings, function (savedSettings) {
		usecachesetting = savedSettings.usecacheforanalysis;
		cacheset.checked = savedSettings.cachepages;
		usecache.checked = cacheset.checked && usecachesetting;
		cachelim.value = savedSettings.pagecachetimelimit;
		analimit.value = savedSettings.analysiscachetimelimit;
		pagemodi.checked = savedSettings.populatepages;
		onpageac.checked = savedSettings.onpageactions;
		autorefr.value = savedSettings.autorefreshevery;
		settings = savedSettings;

		if (!cacheset.checked) {
			usecaclb.classList.add("disabled");
			usecache.disabled = true;
		}

		cacheset.addEventListener("change", saved_status);
		usecache.addEventListener("change", saved_status);
		cachelim.addEventListener("change", saved_status);
		analimit.addEventListener("change", saved_status);
		pagemodi.addEventListener("change", saved_status);
		onpageac.addEventListener("change", saved_status);

		statusms.textContent = 'Status: Waiting for changes.';
		document.getElementById('save').disabled = false;
	});
}

function saved_status(e) {
	if (e.srcElement.id == "cachepage") {
		if (!cacheset.checked) {
			usecachesetting = usecache.checked;
			usecache.disabled = true;
			usecache.checked = false;
			usecaclb.classList.add("disabled");
		} else {
			usecache.disabled = false;
			usecache.checked = usecachesetting;
			usecaclb.classList.remove("disabled");
		};
	};

	if (e.srcElement.id == "usecachepages") {
		usecachesetting = usecache.checked; //only change the setting when user actively selects it.
	}

	statusms.textContent = "Status: Changed settings, not yet saved."
};

document.addEventListener('DOMContentLoaded', restore_options);
document.getElementById('save').addEventListener('click', save_options);


popup.js
// NeoGAF Thread Summarizer Copyright (c) 2015 hesido.com
"use strict";

document.addEventListener('DOMContentLoaded', function () {
	var activeTab = false,
	settings,
	//threadExtract,
	threadId,
	//thread,
	threadURL,
	threadCurPage,
	threadLastPage,
	threadTitle,
	command1 = document.getElementById("analyze"),
	command2 = document.getElementById("abort"),
	command3 = document.getElementById("show"),
	command4 = document.getElementById("clearcache"),
	message1 = document.getElementById("message"),
	coverdiv = document.getElementById("quotedisplaysettings"),
	inputvl1 = document.getElementById("threshold"),
	ordersel = document.getElementById("ordertype"),
	cacheset = document.getElementById("cachepage"),
	poplpage = document.getElementById("populpage"),
	follwusr = document.getElementById("followuser"),
	autorefr = document.getElementById("autorefreshlastpage"),
	//progress = document.getElementById("progress"),
	//progrbar = document.getElementById("bar"),
	setbuttn = document.getElementById("settingsbutton"),
	settndiv = document.getElementById("generalsettings"),
	refreshp = document.getElementById("autorefreshtext"),
	baranima = new $fleXanim.Prepare();

	baranima.setTemplate({
		template : "width:##%",
		values : [0]
	}).clearAnim().animate("bar"); //set template

	chrome.runtime.sendMessage(null, {
		action : "sendSettings"
	});

	chrome.tabs.query({
		currentWindow : true,
		active : true
	}, function (tabs) {
		activeTab = tabs[0];

		//chrome.tabs.executeScript(activeTab.id, {
		//	file : 'analyzepage.js'
		//});

		chrome.tabs.sendMessage(activeTab.id, {
			action : "pageInfoAnalyse",
		});

	});

	setbuttn.addEventListener('click', function () {
		settndiv.classList.toggle("activated");
	});

	chrome.runtime.onMessage.addListener(function (request, sender) {

		if (request.action == "popupUIcommand_applysettings") {

			settings = request.settings;
			inputvl1.value = settings.threshold;
			ordersel.selectedIndex = settings.ordertype;
			follwusr.value = settings.followuser;
			cacheset.checked = settings.cachepages;
			poplpage.checked = settings.populatepages;
			autorefr.value = settings.autorefreshevery;
			inputvl1.addEventListener("change", submitSetting);
			ordersel.addEventListener("change", submitSetting);
			follwusr.addEventListener("change", submitSetting);
			cacheset.addEventListener("change", submitSetting);
			poplpage.addEventListener("change", submitSetting);
			autorefr.addEventListener("change", submitSetting);
			setVisibility();
		};

		request.targetTab = request.targetTab || (sender.tab && sender.tab.id) || false;

		if (activeTab.id !== request.targetTab) //only display responses for that of the thread in the active tab.
			return;

		threadId = threadId || request.threadId || 0;

		if (request.action == "popupUIcommand_setThreadInfo") {
			//activate only when thread info is received. //the info can be shorthanded as request.
			threadURL = request.url;
			threadCurPage = request.curPage;
			threadLastPage = request.lastPage;
			threadTitle = request.threadTitle;
			
			command1.disabled = false;
			command1.addEventListener('click', function () {
				if (threadURL) {
					chrome.runtime.sendMessage(null, {
						action : "startAnalyze",
						threadId : threadId,
						url : threadURL,
						tabId : activeTab.id,
						curPage : threadCurPage,
						lastPage: threadLastPage,
						threadTitle : threadTitle
					});
				} else {
					message1.textContent = "Error: Cannot extract thread information."
				};

			});

			command2.addEventListener('click', function () {

				chrome.runtime.sendMessage(null, {
					action : "abortAnalyze",
					threadId : threadId,
					tabId : activeTab.id,
				});
				command1.disabled = false;
				command2.disabled = true;
				message1.textContent = "Aborted analysis, waiting for processes to finish";
				progressBar(0, 40);

			});

			command3.addEventListener('click', function () {
				chrome.runtime.sendMessage(null, {
					action : "displayResults",
					threadId : threadId,
					//threshold : inputvl1.value,
					//listusers : follwusr.value,
					//sortByQuoteNo : ordersel.selectedIndex,
					tabId : activeTab.id
				});
			});

			command4.addEventListener('click', function () {
				chrome.runtime.sendMessage(null, {
					action : "clearThreadData",
					threadId : threadId,
					tabId : activeTab.id,
				});
				coverdiv.classList.remove("activated");
				command1.disabled = false;
				command3.disabled = true;
			});

		};

		if (request.action == "popupUIcommand_displayThreadStatus") {

			message1.textContent = "Thread status: " + request.status;
			if (request.status == "Analysis in progress") {
				command2.disabled = false;
			};
			if (request.status == "Analysis completed") {
				coverdiv.classList.add("activated");
				command2.disabled = true;
				command3.disabled = false;
				command4.disabled = false;
				progressBar(100, 10);
			};

			if (request.status == "Cleared thread cache") {
				message1.textContent = request.status;
				progressBar(0, 20);
				command1.disabled = false;
				command3.disabled = true;
				coverdiv.classList.remove("activated");
			};

		};

		if (request.action == "popupUIcommand_analyzing") {

			command1.disabled = true;
			command2.disabled = false;
			message1.textContent = "Analysis in progress";
		};

		if (request.action == "popupUIcommand_displayProgress") {
			var percentage = parseInt(((request.pageNo - request.analyzeStartPage + 1) / (request.lastPage - request.analyzeStartPage + 1)) * 100);
			if (request.status !== "Analysis in progress")
				return;
			progressBar(percentage);
		};

		if (request.action == "popupUIcommand_analyzedAlert") {

			command2.disabled = true;
			progressBar(100, 15);
			message1.textContent = request.message;
		};

		if (request.action == "popupUIcommand_analyzeComplete") {

			command2.disabled = true;
			coverdiv.classList.add("activated");
			command3.disabled = false;
			command4.disabled = false;
			message1.textContent = "Analysis completed";
			progressBar(100, 15);
		};

		if (request.action == "popupUIcommand_error") {

			message1.textContent = request.message;
		};

		if (request.action == "popupUIcommand_message") {

			message1.textContent = request.message;
		};

	});
	
	function setVisibility() {
		if(!settings.cachepages) refreshp.classList.add("disabled");
		else refreshp.classList.remove("disabled");
	}

	function submitSetting() {
		settings.threshold = inputvl1.value;
		settings.ordertype = ordersel.selectedIndex;
		settings.followuser = follwusr.value;
		settings.cachepages = cacheset.checked;
		settings.populatepages = poplpage.checked;
		settings.autorefreshevery = autorefr.value;
		setVisibility();

		chrome.runtime.sendMessage(null, {
			action : "receiveSettings",
			settings : settings,
			tabId : activeTab.id
		});

	};

	function progressBar(percentage, frameLength) {
		baranima.reset().setAnimation({
			template : "width:##%",
			endVal : [percentage],
			frames : frameLength || 8,
			tween : "easeout"
		}).clearAnim().animate("bar");
	};

});

//fleXanim v1.1 beta 5 //minimal animation library by hesido.com
var $fleXanim = {
	aeT : {},
	Prepare : function () {
		this.aeL = {};
		this.aQ = []
	}
};
$fleXanim.Prepare.prototype = {
	setAnimation: function (s) {
		if (!s.cachE)
			s.cachE = {
				aP: s.template.split("##"),
				fP: [],
				fr: s.frames || 25,
				ms: s.milisecs || 16,
				tw: s.tween || "smooth"
			};
		var cV = s.cachE;
		for (var i = 1, reg = /^\.#/, aP, tT; aP = cV.aP[i]; i++) {
			tT = reg.test(aP);
			cV.fP[i - 1] = tT;
			if (tT)
				cV.aP[i] = aP.slice(2)
		};

		this.aQ.push(s);
		return this
	},
	setTemplate: function (s) {
		var aP = s.template.split("##");
		for (var i = 0, reg = /^\.#/, propS = '', rP; aP[i]; i++) {
			rP = s.values[i] || 0;
			propS += aP[i].replace(reg, '') + rP
		};
		s.stylE = propS;
		this.aQ.push(s);
		return this
	},
	run: function (f) {
		this.aQ.push({
			func: f
		});
		return this
	},
	forward: function (reF, pD) {
		var aE = this.aeL[reF];
		if (!aE)
			return;
		aE.pD = pD || 1;
		if (!aE.animInt)
			this.runAnim(aE.paused || (pD && pD < 0 && this.aQ.length - 1) || 0, reF);
		aE.paused = false
	},
	reverse: function (reF) {
		this.forward(reF, -1)
	},
	animate: function (reF) {
		if (!this.aeL[reF])
			this.init(reF);
		this.forward(reF)
	},
	init: function (reF) {
		this.aeL[reF] = {
			aI: 0,
			sV: [],
			eV: [],
			ref: reF,
			elm: document.getElementById(reF)
		};
		$fleXanim.aeT[reF] = $fleXanim.aeT[reF] || {};
		return this.aeL[reF]
	},
	clearAnim: function () {
		this.aQ.push({
			clear: true
		});
		return this
	},
	delay: function (dA) {
		this.aQ.push({
			delay: dA
		});
		return this
	},
	loopBegin: function () {
		this.loopStart = this.aQ.length;
		return this
	},
	loop: function (n) {
		this.aQ.push({
			doLoop: true,
			loopTimes: n || -1,
			loopStart: this.loopStart || 0
		});
		return this
	},
	setStyle: function (p) {
		this.aQ.push({
			stylE: p
		});
		return this
	},
	breakLoop: function (reF) {
		this.aeL[reF].breakOut = true
	},
	reset: function () {
		this.pause(false, true);
		this.aQ = [];
		this.aeL = {};
		return this
	},
	pause: function (reF, reset) {
		if (!reF) {
			for (var i in this.aeL) {
				if (this.aeL[i].sV)
					this.pause(i, reset)
			};
			return
		};
		var aE = this.aeL[reF];
		if (!aE)
			return;
		window.clearInterval(aE.animInt);
		window.clearTimeout(aE.animInt);
		aE.animInt = false;
		if (!reset)
			aE.paused = aE.aI
	},
	applyStyle: function (p, elm) {
		var sL = p.match(/[^:;]+/g);
		for (var i = 0, proP, vaL; proP = sL[i]; i += 2) {
			vaL = sL[i + 1];
			elm.style[proP] = vaL
		}
	},
	runAnim: function (inD, reF) {
		var aO = this,
		aE = this.aeL[reF],
		s,
		aV,
		fW = (aE.pD > 0),
		svH,
		evH;
		aE.aI = inD;
		aE.animRun = true;
		if (!(s = aO.aQ[inD]))
			return;
		if (aE.animInt)
			return;
		if (s.clear) { this.aeL[reF] = null; delete this.aeL[reF]; return; }
		if (s.func)
			s.func(aE);
		if (s.doLoop) {
			if (!aE.loopHub)
				aE.loopCount = fW ? 0 : s.loopTimes + 1;
			aE.loopCount += aE.pD;
			if ((s.loopTimes == -1 || (aE.loopCount <= s.loopTimes && aE.loopCount >= 0)) && !aE.breakOut) {
				aE.loopHub = inD;
				aE.loopStart = s.loopStart;
				aO.runAnim(fW ? aE.loopStart : inD - 1, reF);
				return
			} else {
				aE.loopHub = aE.breakOut = false;
				aO.runAnim(fW ? inD + 1 : aE.loopStart - 1, reF);
				return
			}
		};
		if (s.stylE)
			aO.applyStyle(s.stylE, aE.elm);
		if (s.values)
			$fleXanim.aeT[reF][s.template] = s.values.slice(0);
		if (s.delay) {
			aE.animInt = window.setTimeout(function () {
				aE.animInt = false;
				aO.runAnim(inD + aE.pD, reF)
			}, s.delay);
			return
		};
		if (!(aV = s.cachE)) {
			aO.runAnim(inD + aE.pD, reF);
			return
		};
		aE.cF = aE.cF || (fW ? 0 : aV.fr);
		aE.currentVal = [];
		if (!aE.sV[inD])
			aE.sV[inD] = (s.startVal && s.startVal.slice(0)) || $fleXanim.aeT[reF][s.template] || [];
		svH = aE.sV[inD];
		if (!aE.eV[inD])
			aE.eV[inD] = (s.endVal && s.endVal.slice(0)) || [];
		evH = aE.eV[inD];
		for (var i = 0, j = (s.relVal && s.relVal.length) || 0, rV; i < j; i++) {
			rV = s.relVal[i];
			evH[i] = (svH[i] || 0) + rV

		};

		aE.animInt = window.requestAnimationFrame(executeAnim);
		function executeAnim() {
			for (var i = 0, propS = '', rP; aV.aP[i]; i++) {
				rP = ((svH[i] || svH[i] === 0) && (evH[i] || evH[i] === 0) && aO.animStep[aV.tw](aE.cF, svH[i], evH[i], aV.fr));
				rP = (rP || rP === 0) ? rP : ""; //added in 1.1 beta - a more elegant and shorter solution is welcome.
				rP = (rP && aV.fP[i] === false) ? parseInt(rP) : rP;
				propS += aV.aP[i] + rP;
				aE.currentVal[i] = rP
			};
			aO.applyStyle(propS, aE.elm);
			$fleXanim.aeT[reF][s.template] = aE.currentVal.slice(0);
			if (s.run)
				s.run(aE);
			aE.cF += aE.pD;
			if (aE.cF > aV.fr || aE.cF < 0) {
				aE.cF = aE.animInt = false;
				delete aE.currentVal;
				aE.aI += aE.pD;
				if (aE.loopHub && aE.pD < 0 && aE.aI < aE.loopStart)
					aE.aI = aE.loopHub;
				if (aO.aQ[aE.aI])
					aO.runAnim(aE.aI, reF);
				else
					aE.animRun = false
			} else
				aE.animInt = window.requestAnimationFrame(executeAnim)
		}
	},
	animStep: {
		smooth: function (t, b, c, d) {
			c -= b;
			var ts = (t /= d) * t,
			tc = ts * t;
			return (b + c * (-2 * tc + 3 * ts))
		},
		easein: function (t, b, c, d) {
			c -= b;
			var ts = (t /= d) * t * t;
			return (b + c * (ts))
		},
		easeout: function (t, b, c, d) {
			c -= b;
			var ts = (t /= d) * t,
			tc = ts * t;
			return (b + c * (tc - 3 * ts + 3 * t))
		},
		linear: function (t, b, c, d) {
			c -= b;
			t /= d;
			return (b + c * (t))
		}
	}
};
